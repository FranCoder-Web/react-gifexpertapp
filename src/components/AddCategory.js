import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategories }) => {
	const [inputValue, setInputValue] = useState('')

	const handleInputChange = (e) => {
		setInputValue(e.target.value)
	}
	const handleSubmit = (e) => {
		e.preventDefault()

		if (inputValue.trim().length > 2) {
			setCategories((prevState) => {
				let repeatCategorie = prevState.indexOf(
					inputValue.toUpperCase().trim()
				)
				if (repeatCategorie >= 0) {
					return prevState
				} else {
					return [inputValue.toUpperCase(), ...prevState]
				}
			})
			setInputValue('')
		}
	}
	return (
		<form onSubmit={handleSubmit}>
			<input
				className='search'
				type='text'
				value={inputValue}
				onChange={handleInputChange}
				placeholder='SEARCH GIFS...'
			/>
		</form>
	)
}

AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired,
}
