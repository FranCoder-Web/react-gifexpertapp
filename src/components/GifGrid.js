import React from 'react'
import PropTypes from 'prop-types'
import { GifGridItem } from './GifGridItem'
import { useFetchGifs } from '../hooks/useFetchGifs'
import Loader from 'react-loader-spinner'

export const GifGrid = ({ category, cleanCategory }) => {
	const { data: images, loading } = useFetchGifs(category)
	const handleCategory = () => {
		cleanCategory((prevstate) => {
			return prevstate.filter((cat) => cat !== category)
		})
	}

	return (
		<>
			<h2 className='category'>
				{category}
				<span onClick={handleCategory} className='x-category'>
					X
				</span>
			</h2>
			{loading && (
				<Loader
					type='ThreeDots'
					color='#1e272e'
					height={120}
					width={120}
				/>
			)}
			<div className='card-grid'>
				{images.map((img) => (
					<GifGridItem key={img.id} category={category} {...img} />
				))}
			</div>
		</>
	)
}

GifGrid.propTypes = {
	category: PropTypes.string.isRequired,
}
