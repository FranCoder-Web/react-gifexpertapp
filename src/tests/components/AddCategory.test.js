import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import { AddCategory } from '../../components/AddCategory'

describe('Probando el componente <AddCategory />', () => {
	const setCategories = jest.fn()
	let wrapper = shallow(<AddCategory setCategories={setCategories} />)

	beforeEach(() => {
		jest.clearAllMocks()
		wrapper = shallow(<AddCategory setCategories={setCategories} />)
	})

	test('debe mostrarse correctamente', () => {
		expect(wrapper).toMatchSnapshot()
	})

	test('debe cambiar la caja del input', () => {
		const input = wrapper.find('input')
		const value = 'Hola mundo'

		input.simulate('change', { target: { value } })
	})

	test('NO debe postear la informacion con el submit', () => {
		wrapper.find('form').simulate('submit', { preventDefault: () => {} })

		expect(setCategories).not.toHaveBeenCalled()
	})

	test('debe llamar a setCategories y limpiar la caja de texto', () => {
		const input = wrapper.find('input')
		const value = 'Hola mundo'
		input.simulate('change', { target: { value } })

		wrapper.find('form').simulate('submit', { preventDefault: () => {} })
		expect(setCategories).toHaveBeenCalled()
		expect(setCategories).toHaveBeenCalledWith(expect.any(Function))

		expect(input.prop('value')).toBe('')
	})
})
