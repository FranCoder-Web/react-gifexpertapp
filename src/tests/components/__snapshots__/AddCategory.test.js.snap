// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente <AddCategory /> debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    className="search"
    onChange={[Function]}
    placeholder="SEARCH GIFS..."
    type="text"
    value=""
  />
</form>
`;
